File: backend/src/models/Group.js
const { getDb } = require('../config/firebase');
const { serializeDoc } = require('../utils/firestore');

const COLLECTION_NAME = 'groups';

// Generate invite code
const generateInviteCode = () => {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    let code = '';
    for (let i = 0; i < 6; i++) {
        code += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return code;
};

// Create group
const create = async (groupData) => {
    const db = getDb();

    // Validate members
    if (!groupData.members || !Array.isArray(groupData.members)) {
        throw new Error('Invalid members data');
    }

    const group = {
        name: groupData.name?.trim() || '',
        description: groupData.description?.trim() || '',
        members: groupData.members || [],
        expenses: [],
        createdBy: groupData.createdBy,
        inviteCode: generateInviteCode(),
        imageUrl: groupData.imageUrl || '',
        createdAt: new Date(),
        updatedAt: new Date()
    };

    const docRef = await db.collection(COLLECTION_NAME).add(group);
    return { id: docRef.id, ...group, createdAt: group.createdAt.toISOString(), updatedAt: group.updatedAt.toISOString() };
};

// Find groups by member userId
const findByMember = async (userId) => {
    const db = getDb();
    // Fetch all groups and filter (Firestore restriction workaround)
    const snapshot = await db.collection(COLLECTION_NAME).get();

    return snapshot.docs
        .map(doc => serializeDoc(doc))
        .filter(group => {
            if (!group || !Array.isArray(group.members)) return false;
            return group.members.some(member => member.userId === userId);
        });
};

// Find group by ID
const findById = async (id) => {
    const db = getDb();
    const doc = await db.collection(COLLECTION_NAME).doc(id).get();

    if (!doc.exists) return null;
    return serializeDoc(doc);
};

// Find group by invite code
const findByInviteCode = async (code) => {
    const db = getDb();
    const snapshot = await db.collection(COLLECTION_NAME)
        .where('inviteCode', '==', code.toUpperCase())
        .limit(1)
        .get();

    if (snapshot.empty) return null;
    return serializeDoc(snapshot.docs[0]);
};

// Update group
const updateById = async (id, updateData) => {
    const db = getDb();
    updateData.updatedAt = new Date();
    await db.collection(COLLECTION_NAME).doc(id).update(updateData);
    return await findById(id);
};

// Delete group
const deleteById = async (id) => {
    const db = getDb();
    await db.collection(COLLECTION_NAME).doc(id).delete();
    return true;
};

// Add member to group
const addMember = async (groupId, member) => {
    const group = await findById(groupId);
    if (!group) return null;

    // Check if already a member
    if (group.members.some(m => m.userId === member.userId)) {
        return group;
    }

    const newMember = {
        userId: member.userId,
        name: member.name,
        email: member.email,
        upiId: member.upiId || '',
        amountOwed: 0,
        amountLent: 0
    };

    group.members.push(newMember);
    return await updateById(groupId, { members: group.members });
};

// Remove member from group
const removeMember = async (groupId, userId) => {
    const group = await findById(groupId);
    if (!group) return null;

    group.members = group.members.filter(m => m.userId !== userId);
    return await updateById(groupId, { members: group.members });
};

// Add expense to group
const addExpense = async (groupId, expense) => {
    const group = await findById(groupId);
    if (!group) return null;

    const newExpense = {
        id: `exp_${Date.now()}`,
        paidBy: expense.paidBy,
        paidByName: expense.paidByName,
        amount: expense.amount,
        description: expense.description || 'Group Expense',
        splits: expense.splits || [],
        category: expense.category || 'other',
        date: expense.date ? new Date(expense.date) : new Date(),
        createdAt: new Date()
    };

    group.expenses.push(newExpense);

    // Update member balances logic 
    const payer = group.members.find(m => m.userId === expense.paidBy);
    if (payer) {
        let totalSplit = 0;
        expense.splits.forEach(split => {
            const member = group.members.find(m => m.userId === split.memberId);
            if (member && member.userId !== expense.paidBy) {
                member.amountOwed += split.amount;
                totalSplit += split.amount;
            }
        });
        payer.amountLent += totalSplit;
    }

    return await updateById(groupId, {
        expenses: group.expenses,
        members: group.members
    });
};

// Get total expenses for group
const getTotalExpenses = (group) => {
    return group.expenses.reduce((sum, expense) => sum + expense.amount, 0);
};

// Get member balance
const getMemberBalance = (group, userId) => {
    const member = group.members.find(m => m.userId === userId);
    if (!member) return 0;
    return member.amountLent - member.amountOwed;
};

module.exports = {
    COLLECTION_NAME,
    generateInviteCode,
    create,
    findByMember,
    findById,
    findByInviteCode,
    updateById,
    deleteById,
    addMember,
    removeMember,
    addExpense,
    getTotalExpenses,
    getMemberBalance
};

File: mobile/lib/models/group.dart
class GroupMember {
  final String userId;
  final String name;
  final String email;
  final String upiId; // Member's UPI ID for settlements (optional)
  final double amountOwed; // How much this user owes in total
  final double amountLent; // How much this user has lent

  GroupMember({
    required this.userId,
    required this.name,
    required this.email,
    this.upiId = '',
    this.amountOwed = 0.0,
    this.amountLent = 0.0,
  });

  factory GroupMember.fromJson(Map<String, dynamic> json) {
    return GroupMember(
      userId: json['userId'] ?? '',
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      upiId: json['upiId'] ?? '',
      amountOwed: (json['amountOwed'] ?? 0).toDouble(),
      amountLent: (json['amountLent'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'name': name,
      'email': email,
      'upiId': upiId,
      'amountOwed': amountOwed,
      'amountLent': amountLent,
    };
  }

  double get balance => amountLent - amountOwed;
}

class GroupExpenseSplit {
  final String memberId;
  final String memberName;
  final double amount;

  GroupExpenseSplit({
    required this.memberId,
    required this.memberName,
    required this.amount,
  });

  factory GroupExpenseSplit.fromJson(Map<String, dynamic> json) {
    return GroupExpenseSplit(
      memberId: json['memberId'] ?? '',
      memberName: json['memberName'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'memberId': memberId,
      'memberName': memberName,
      'amount': amount,
    };
  }
}

class GroupExpense {
  final String id;
  final String groupId;
  final String paidBy; // userId
  final String paidByName;
  final double amount;
  final String description;
  final List<GroupExpenseSplit> splits;
  final DateTime date;
  final String category;

  GroupExpense({
    required this.id,
    required this.groupId,
    required this.paidBy,
    required this.paidByName,
    required this.amount,
    required this.description,
    required this.splits,
    required this.date,
    required this.category,
  });

  factory GroupExpense.fromJson(Map<String, dynamic> json) {
    return GroupExpense(
      id: json['_id'] ?? json['id'] ?? '',  // MongoDB uses _id
      groupId: json['groupId'] ?? '',
      paidBy: json['paidBy'] ?? '',
      paidByName: json['paidByName'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      description: json['description'] ?? '',
      splits: (json['splits'] as List<dynamic>?)
          ?.map((e) => GroupExpenseSplit.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
      category: json['category'] ?? 'others',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'groupId': groupId,
      'paidBy': paidBy,
      'paidByName': paidByName,
      'amount': amount,
      'description': description,
      'splits': splits.map((e) => e.toJson()).toList(),
      'date': date.toIso8601String(),
      'category': category,
    };
  }
}

class Group {
  final String id;
  final String name;
  final String description;
  final List<GroupMember> members;
  final List<GroupExpense> expenses;
  final DateTime createdAt;
  final String createdBy;
  final String imageUrl;
  final String inviteCode;

  Group({
    required this.id,
    required this.name,
    required this.description,
    required this.members,
    required this.expenses,
    required this.createdAt,
    required this.createdBy,
    this.imageUrl = '',
    this.inviteCode = '',
  });

  factory Group.fromJson(Map<String, dynamic> json) {
    return Group(
      id: json['_id'] ?? json['id'] ?? '',  // MongoDB uses _id
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      members: (json['members'] as List<dynamic>?)
          ?.map((e) => GroupMember.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      expenses: (json['expenses'] as List<dynamic>?)
          ?.map((e) => GroupExpense.fromJson(e as Map<String, dynamic>))
          .toList() ?? [],
      createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
      createdBy: json['createdBy'] ?? '',
      imageUrl: json['imageUrl'] ?? '',
      inviteCode: json['inviteCode'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'members': members.map((e) => e.toJson()).toList(),
      'expenses': expenses.map((e) => e.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'createdBy': createdBy,
      'imageUrl': imageUrl,
      'inviteCode': inviteCode,
    };
  }

  double getTotalExpenses() {
    return expenses.fold(0, (sum, expense) => sum + expense.amount);
  }

  Map<String, double> getSettlements() {
    Map<String, double> settlements = {};
    
    for (var member in members) {
      settlements[member.userId] = member.balance;
    }
    
    return settlements;
  }
}


